### PR TITLE
KAFKA-8531: Change default replication factor config

### PR DESCRIPTION
**Pull request by mjsax**
Call for review @cadonna @ableegoldman

### CODE REVIEW

**Comment by mjsax** [comment]
As test against older broker versions, I started https://jenkins.confluent.io/job/system-test-kafka-branch-builder/4463/ -- maybe we need to update some system tests...

**Comment by mjsax** [comment]
Was also wondering about a potential error message -- not sure atm what error message a user would get if they run against 2.3 brokers and if the error message would be clear. Should we do anything about it?

**Comment by ableegoldman** [comment]
> Was also wondering about a potential error message -- not sure atm what error message a user would get if they run against 2.3 brokers and if the error message would be clear. Should we do anything about it?

Not sure either -- maybe you can use the soak test to spin up brokers on 2.2 against this PR and check out the error message + stack trace? I definitely think we should try to catch the error and log a more helpful error message (eg `in 3.0 we changed the default replication factor to -1 which requires brokers to be on 2.3+, please update your brokers or manually set the replication factor config`)

**Comment by mjsax** [comment]
The logs show the following error:
```
[2021-04-13 15:52:17,685] ERROR stream-thread [main] Unexpected error during topic creation for streams-wordcount-KSTREAM-AGGREGATE-STATE-STORE-0000000003-repartition.
Error message was: org.apache.kafka.common.errors.UnsupportedVersionException: Creating topics with default partitions/replication factor are only supported in CreateTopicRequest version 4+. The following topics need values for partitions and replicas: [streams-wordcount-KSTREAM-AGGREGATE-STATE-STORE-0000000003-repartition] (org.apache.kafka.streams.processor.internals.InternalTopicManager:446)
[2021-04-13 15:52:17,687] ERROR stream-client [streams-wordcount-4b9ea556-bc49-4671-83e8-4d2adaba3677] Encountered the following exception during processing and the registered exception handler opted to SHUTDOWN_CLIENT. The streams client is going to shut down now.  (org.apache.kafka.streams.KafkaStreams:484)
org.apache.kafka.streams.errors.StreamsException: Could not create topic streams-wordcount-KSTREAM-AGGREGATE-STATE-STORE-0000000003-repartition.
...
Caused by: org.apache.kafka.common.errors.UnsupportedVersionException: Creating topics with default partitions/replication factor are only supported in CreateTopicRequest version 4+. The following topics need values for partitions and replicas: [streams-wordcount-KSTREAM-AGGREGATE-STATE-STORE-0000000003-repartition]

```

Do we think this would be sufficient? Or should we try to handle this specific `UnsupportedVersionException` explicitly?

**Comment by ableegoldman** [comment]
Yikes -- no, I think we definitely need to handle that ourselves. I don't think users will have any idea what that means -- for one thing they'll wonder why they're suddenly seeing it now, and probably report it as a bug if we don't explicitly tell them that we changed the default replication factor. For another, how are they supposed to know how to interpret `Creating topics with default partitions/replication factor are only supported in CreateTopicRequest version 4+` -- it doesn't even clearly state that the problem is older brokers, much less what version specifically the brokers need to be

**Comment by mjsax** [comment]
Updated the PR.

The log now show:
```
[2021-04-14 14:10:02,176] ERROR stream-thread [main] Unexpected error during topic creation for streams-wordcount-KSTREAM-AGGREGATE-STATE-STORE-0000000003-repartition.
Error message was: org.apache.kafka.common.errors.UnsupportedVersionException: Creating topics with default partitions/replication factor are only supported in CreateTopicRequest version 4+. The following topics need values for partitions and replicas: [streams-wordcount-KSTREAM-AGGREGATE-STATE-STORE-0000000003-repartition] (org.apache.kafka.streams.processor.internals.InternalTopicManager:451)
[2021-04-14 14:10:02,183] ERROR stream-client [streams-wordcount-4b9ea556-bc49-4671-83e8-4d2adaba3677] Encountered the following exception during processing and the registered exception handler opted to SHUTDOWN_CLIENT. The streams client is going to shut down now.  (org.apache.kafka.streams.KafkaStreams:484)
org.apache.kafka.streams.errors.StreamsException: Could not create topic streams-wordcount-KSTREAM-AGGREGATE-STATE-STORE-0000000003-repartition, because brokers don't support configuration replication.factor=-1. You can change the replication.factor config or upgrade your brokers to version 2.4 or newer to avoid this error.
	at org.apache.kafka.streams.processor.internals.InternalTopicManager.makeReady(InternalTopicManager.java:459)
[...]
```

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java [hunk 1]
```diff
@@ -16,9 +16,11 @@
  */
 package org.apache.kafka.streams.processor.internals;
 
+import java.util.Collection;
```
Comment (ableegoldman) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java]: nit: import in the wrong place

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java [hunk 2]
```diff
@@ -439,13 +440,34 @@ private String getBrokerSideConfigValue(final Config brokerSideTopicConfig,
...
                             log.error("Unexpected error during topic creation for {}.\n" +
                                 "Error message was: {}", topicName, cause.toString());
-                            throw new StreamsException(String.format("Could not create topic %s.", topicName), cause);
+
+                            if (cause instanceof UnsupportedVersionException) {
+                                final String errorMessage = cause.getMessage();
+                                if (errorMessage != null &&
+                                    errorMessage.startsWith("Creating topics with default partitions/replication factor are only supported in CreateTopicRequest version 4+")) {
+
+                                    throw new StreamsException(String.format(
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:459]: nit: can we also log an error before throwing, possibly with the actual `cause` message

**Review by ableegoldman — APPROVED** [review]
Two nits, otherwise LGTM. Thanks for adding a more informative error

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java (hunk 2 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:459]: Don't we log the error in L450 above already?

Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:459]: Ah, I was thinking of a log specific to this error but if we just log it at all that's good enough for me. Nevermind this

**Comment by mjsax** [comment]
I started a quick KIP for this change -- will wait until the KIP is accepted before merging.

**Pull request merged**